test/fork/basic-functionality-tests.js

```diff
@@ -1,331 +1,560 @@
+    }
+    expect(currentTotalAssetValue).to.be.greaterThan(mcrEth);
+  });
 
-    await submitGovernanceProposal(ProposalCategory.upgradeNonProxy, upgradesActionDataNonProxy, voters, governance);
+  it('sell NXM Token', async function () {
+    const sellValue = parseEther('1');
+    const buyer = this.abMembers[0];
+    const buyerAddress = buyer.getAddress();
 
-    const tfStoredAddress = await master.getLatestAddress(hex('TF'));
-    assert.equal(tfStoredAddress, tokenFunctionsImplementation.address);
+    const balanceBefore = await ethers.provider.getBalance(buyerAddress);
+    const currentTotalAssetValue = await this.pool.getPoolValueInEth();
+    const mcr = await this.mcr.getMCR();
+    const expectedTokensReceived = await this.pool.calculateEthForNXM(sellValue, currentTotalAssetValue, mcr);
 
-    const gwProxy = await OwnedUpgradeabilityProxy.at(await master.getLatestAddress(hex('GW')));
-    const gwImplementation = await gwProxy.implementation();
+    const tx = await this.pool.connect(buyer).sellNXM(sellValue, '0');
+    const receipt = await tx.wait();
+    const txCost = receipt.gasUsed.mul(receipt.effectiveGasPrice);
+    const balanceAfter = await ethers.provider.getBalance(buyerAddress);
 
-    assert.equal(gwImplementation, gatewayImplementation.address);
+    expect(balanceAfter).to.be.equal(balanceBefore.add(expectedTokensReceived).sub(txCost));
   });
 
-  it('performs hypothetical future master upgrade', async function () {
-    const { voters, governance, master } = this;
-
-    const masterProxy = await OwnedUpgradeabilityProxy.at(master.address);
-
-    // upgrade to new master
-    const masterImplementation = await NXMaster.new();
-
-    // vote and upgrade
-    const upgradeMaster = web3.eth.abi.encodeParameters(['address'], [masterImplementation.address]);
-    await submitGovernanceProposal(ProposalCategory.upgradeMaster, upgradeMaster, voters, governance);
+  it('add product types', async function () {
+    const productTypes = [
+      {
+        productTypeName: 'x',
+        productTypeId: MaxUint256,
+        ipfsMetadata: 'protocolCoverIPFSHash',
+        productType: {
+          descriptionIpfsHash: 'protocolCoverIPFSHash',
+          claimMethod: 0,
+          gracePeriod: 30,
+        },
+      },
+      {
+        productTypeName: 'y',
+        productTypeId: MaxUint256,
+        ipfsMetadata: 'custodyCoverIPFSHash',
+        productType: {
+          descriptionIpfsHash: 'custodyCoverIPFSHash',
+          claimMethod: 0,
+          gracePeriod: 90,
+        },
+      },
+      {
+        productTypeName: 'z',
+        productTypeId: MaxUint256,
+        ipfsMetadata: 'yieldTokenCoverIPFSHash',
+        productType: {
+          descriptionIpfsHash: 'yieldTokenCoverIPFSHash',
+          claimMethod: 1,
+          gracePeriod: 14,
+        },
+      },
+    ];
+    try {
+      await this.cover.connect(this.abMembers[0]).setProductTypes(productTypes);
+    } catch (ex) {
+      console.log(ex);
+    }
```

Comment:
let's assert at least the new product type count here to be increased by .length of that one.

Reply:
This was added so I can test the product adding because it was not implemented yet, but yeah we can leave it so we can test this also